backend/src/impl/benchmark_utils.py

```diff
@@ -82,11 +86,27 @@ def generate_dataframe_from_sys_infos(config: BenchmarkConfig, systems: list[dic
         #                interchangeably due to OpenAPI deserialization being
         #                incomplete. Should be fixed.
 
+        # --- Collect each dataset to be included in the benchmark
+        if config.datasets:
+            dataset_to_id = {
+                (
+                    x["dataset_name"],
+                    x.get("sub_dataset", None),
+                    x.get("split", "test"),
+                ): i
+                for i, x in enumerate(config.datasets)
+            }
+        else:
+            dataset_to_id = {}
+            for sys in systems:
+                data_id = (
+                    sys["dataset_name"],
+                    sys["sub_dataset_name"],
+                    sys["dataset_split"],
+                )
+                dataset_to_id[data_id] = dataset_to_id.get(data_id, len(dataset_to_id))
+
         # --- Rearrange so we have each system's result over each dataset
-        dataset_to_id = {
-            (x["dataset_name"], x.get("sub_dataset", None), x.get("split", "test")): i
-            for i, x in enumerate(config.datasets)
-        }
         system_dataset_results: dict[str, list[dict | None]] = {}
         for sys in systems:
             sys_name = sys["system_name"]
```

Comment:
When config.datasets is None, `[None for _ in config.datasets]` throws an error, which is part of the reason I think global_ner is wigging out (since global_ner.json doesn't have datasets).

Reply:
> I don't think we even need to access config.datasets anymore in the function generate_dataframe_from_sys_infos, because if the config file has a config.datasets, we use it to query from the database which returns all systems matching the query. If we don't have a config.datasets in the config file, we use the task name as the query and return all systems matching the query. Either way, the systems input to the generate_dataframe_from_sys_infos should give us everything we need to know.

That's exactly right! I needed to adjust this so that we got all the info we need from `config.datasets` if it exists, and from the retrieved systems if not.

> Also, there are some other things i don't understand about this implementation of generate_dataframe_from_sys_infos

Good question. This part is somewhat important, because we need to have a default behavior that if a system hasn't submitted it gets a score of zero. Otherwise consider the degenerate case where we have a benchmark with 10 datasets, and a team decides to submit a system for only the easiest dataset. Then if we took "mean" we would get the score of that easy dataset instead of an average over all the datasets (including zeros). These first lines of code find all the unique datasets and make sure that each system is processed with respect to each of the unique datasets.